compiler/src/main/java/com/squareup/anvil/compiler/InterfaceMergerIr.kt

```diff
@@ -0,0 +1,276 @@
+              "${scopes.joinToString(prefix = "[", postfix = "]") { it.fqName.asString() }} " +
+              "wants to exclude ${excludedClass.fqName}, but the excluded class isn't " +
+              "contributed to the same scope.",
+            classReference = mergeAnnotatedClass,
+          )
+        }
+      }
+      .toList()
+
+    val supertypes = mergeAnnotatedClass.clazz.superTypes()
+    if (excludedClasses.isNotEmpty()) {
+      val intersect = supertypes
+        .map { it.classOrFail.toClassReference(pluginContext) }
+        .flatMap {
+          it.allSuperTypeClassReferences(pluginContext, includeSelf = true)
+        }
+        .intersect(excludedClasses.toSet())
+
+      if (intersect.isNotEmpty()) {
+        throw AnvilCompilationExceptionClassReferenceIr(
+          classReference = mergeAnnotatedClass,
+          message = "${mergeAnnotatedClass.fqName} excludes types that it implements or " +
+            "extends. These types cannot be excluded. Look at all the super types to find these " +
+            "classes: ${intersect.joinToString { it.fqName.asString() }}.",
+        )
+      }
+    }
+
+    val supertypesToAdd = contributesAnnotations
+      .asSequence()
+      .map { it.declaringClass }
+      .filter { clazz ->
+        clazz !in replacedClasses && clazz !in excludedClasses
+      }
+      .plus(
+        findContributedSubcomponentParentInterfaces(
+          clazz = mergeAnnotatedClass,
+          scopes = scopes,
+          pluginContext = pluginContext,
+          moduleFragment = moduleFragment,
+        ),
+      )
+      // Avoids an error for repeated interfaces.
+      .distinct()
+      .map { it.clazz.typeWith() }
+      .toList()
+
+    // Since we are modifying the state of the code here, this does not need to be reflected in
+    // the associated [ClassReferenceIr] which is more of an initial snapshot.
+    mergeAnnotatedClass.clazz.owner.superTypes += supertypesToAdd
+  }
+
+  private fun findContributedSubcomponentParentInterfaces(
+    clazz: ClassReferenceIr,
+    scopes: Collection<ClassReferenceIr>,
+    pluginContext: IrPluginContext,
+    moduleFragment: IrModuleFragment,
+  ): Sequence<ClassReferenceIr> {
+    return classScanner
+      .findContributedClasses(
+        pluginContext = pluginContext,
+        moduleFragment = moduleFragment,
+        annotation = contributesSubcomponentFqName,
+        scope = null,
+        moduleDescriptorFactory = moduleDescriptorFactory,
+      )
+      .filter {
+        it.atLeastOneAnnotation(contributesSubcomponentFqName).single()
+          .parentScope in scopes
+      }
+      .mapNotNull { contributedSubcomponent ->
+        contributedSubcomponent.classId
+          .generatedAnvilSubcomponent(clazz.classId)
+          .createNestedClassId(Name.identifier(PARENT_COMPONENT))
+          .referenceClassOrNull(pluginContext)
+      }
+  }
+}
+
+// TODO is this necessary in IR? Think all supertypes should already be resolved
```

Comment:
I'm checking in kotlin-lang to see if this is still necessary, I suspect it's not

Reply:
Still necessary, it's only direct supertypes by default